From d108632ee2c5cdafcbb6f209e39af913a0a01fce Mon Sep 17 00:00:00 2001
From: Surge Raval <Surge1223@gmail.com>
Date: Wed, 4 Jan 2017 10:29:34 -0800
Subject: [PATCH 3/3] Sepolicy exceptions for font engine assets

Derived from https://substratum.review/#/c/432/
We only want the minimum policy to support our
font engine

Change-Id: I0420de579aed0cff5add181cd0a8bf0f2b05d723
---
 private/app.te           | 4 ++++
 private/file_contexts    | 3 +++
 private/system_app.te    | 3 +++
 private/system_server.te | 3 +++
 private/zygote.te        | 4 ++++
 public/file.te           | 3 +++
 6 files changed, 20 insertions(+)

diff --git a/private/app.te b/private/app.te
index c0deb085..55409f53 100644
--- a/private/app.te
+++ b/private/app.te
@@ -542,3 +542,7 @@ neverallow {
   -bluetooth
   -system_app
 } bluetooth_prop:file create_file_perms;
+
+# Themed resources (i.e. composed icons)
+allow appdomain theme_data_file:dir r_dir_perms;
+allow appdomain theme_data_file:file r_file_perms;
diff --git a/private/file_contexts b/private/file_contexts
index 02e9fd74..f62cf5b0 100644
--- a/private/file_contexts
+++ b/private/file_contexts
@@ -402,6 +402,9 @@
 # Bootchart data
 /data/bootchart(/.*)?		u:object_r:bootchart_data_file:s0
 
+# Themes data
+/data/system/theme(/.*)?    u:object_r:theme_data_file:s0
+
 #############################
 # Expanded data files
 #
diff --git a/private/system_app.te b/private/system_app.te
index 13391f19..1dac70ed 100644
--- a/private/system_app.te
+++ b/private/system_app.te
@@ -87,6 +87,9 @@ allow system_app selinuxfs:file r_file_perms;
 # /sys access
 r_dir_file(system_app, sysfs_type)
 
+# /data/system access
+allow system_app system_data_file:file r_file_perms;
+
 control_logd(system_app)
 read_runtime_log_tags(system_app)
 
diff --git a/private/system_server.te b/private/system_server.te
index 78985346..5d367cca 100644
--- a/private/system_server.te
+++ b/private/system_server.te
@@ -709,6 +709,9 @@ with_asan(`
   allow system_server zygote_exec:file rx_file_perms;
 ')
 
+# allow system_server to look into theme resources
+allow system_server theme_data_file:dir search;
+
 ###
 ### Neverallow rules
 ###
diff --git a/private/zygote.te b/private/zygote.te
index daabbc06..4cd84da9 100644
--- a/private/zygote.te
+++ b/private/zygote.te
@@ -112,6 +112,10 @@ allow zygote tmpfs:dir r_dir_perms;
 # Let the zygote access overlays so it can initialize the AssetManager.
 get_prop(zygote, overlay_prop)
 
+# Themes
+allow zygote theme_data_file:file r_file_perms;
+allow zygote theme_data_file:dir r_dir_perms;
+
 ###
 ### neverallow rules
 ###
diff --git a/public/file.te b/public/file.te
index 1df9f708..0fe4acc4 100644
--- a/public/file.te
+++ b/public/file.te
@@ -346,3 +346,6 @@ with_asan(`type asanwrapper_exec, exec_type, file_type;')
 # Should be:
 #   type apk_data_file, file_type, data_file_type;
 neverallow fs_type file_type:filesystem associate;
+
+# Themes
+type theme_data_file, file_type, data_file_type;
-- 
2.20.0

